app/src/main/java/ee/ioc/phon/android/speechutils/ContinuousRawAudioRecorder.java

```diff
@@ -0,0 +1,215 @@
+            if (e.getMessage() == null) {
+                handleError("Unknown error occurred while initializing recorder");
+            } else {
+                handleError(e.getMessage());
+            }
+        }
+    }
+
+    public ContinuousRawAudioRecorder(int sampleRate) {
+        this(DEFAULT_AUDIO_SOURCE, sampleRate, DEFAULT_BUFFER_LENGTH_IN_MILLIS);
+    }
+
+    public ContinuousRawAudioRecorder() {
+        this(DEFAULT_AUDIO_SOURCE, DEFAULT_SAMPLE_RATE, DEFAULT_BUFFER_LENGTH_IN_MILLIS);
+    }
+
+    public String getWsArgs() {
+        return "?content-type=audio/x-raw,+layout=(string)interleaved,+rate=(int)" + getSampleRate() + ",+format=(string)S16LE,+channels=(int)1";
+    }
+
+    public ContinuousRawAudioRecorder setSessionStartPointer(SessionStartPointer sessionStartPointer) {
+        mSessionStartPointer = sessionStartPointer;
+        return this;
+    }
+
+    private int calculateNumOfSamplesToGoBack(int startPos) {
+        // if the consumed session is not the same as the recorded session
+        // get the data from the beginning of the buffer/desired session start pointer
+        if (!isRecordedSessionSameAsConsumedSession()) {
+            Log.i(LOG_FILTER + "Recorded session and consumed session are NOT the same. Grabbing the data from the session start position");
+            return getNumOfSamplesIn(mSessionStartPointer.getSessionStartPointerMillis());
+        }
+
+        Log.i(LOG_FILTER + "Recorded session and consumed session are the same. Working according to the consumed pointer");
+        if (startPos == getLength()) {
+            Log.i(LOG_FILTER + "Consumed session pointer still points to the recorded session pointer. Nothing to do or to return");
+            return -1;
+        }
+
+        int numOfSamplesToGoBack = getLength() - startPos;
+        if (numOfSamplesToGoBack < 0) // cyclic buffer case
+            numOfSamplesToGoBack += mRecording.length;
+
+        Log.i(LOG_FILTER + "Consumed session pointer is behind the recorded session pointer by: " + numOfSamplesToGoBack + " samples");
+        return numOfSamplesToGoBack;
+    }
+
+    @Override
+    protected byte[] getCurrentRecording(int startPos) {
+
+        int numOfSamplesToGoBack = calculateNumOfSamplesToGoBack(startPos);
+        if (numOfSamplesToGoBack <= 0) {
+            Log.i(LOG_FILTER + "There are no samples that we need to take from the recording");
+            return null;
+        }
+
+        byte[] buffer = new byte[numOfSamplesToGoBack];
+        int potentialStartSample = getLength() - numOfSamplesToGoBack;
+
+        if (potentialStartSample >= 0) {
+            Log.i(LOG_FILTER + "Start sample in the recording is a positive one. Copying from position: " + potentialStartSample + ", " + numOfSamplesToGoBack + " bytes");
+            System.arraycopy(mRecording, potentialStartSample, buffer, 0, numOfSamplesToGoBack);
+        }
+        else {
+            if (!mRecordingBufferIsFullWithData) {
```

Comment:
if the consumption started before the first time that the recording pointer reached the end, the consumption should happen from the beginning of the buffer to now